composer/utils/object_store/gs_object_store.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Google Cloud SDK - Compatible object store."""
+
+from __future__ import annotations
+
+import os
+import pathlib
+import uuid
+from typing import Callable, Optional, Union
+
+from composer.utils.import_helpers import MissingConditionalImportError
+from composer.utils.object_store.object_store import ObjectStore
+
+__all__ = ['GsObjectStore']
+
+BOTOCORE_CLIENT_ERROR_CODES = ('403', '404', 'NoSuchKey')
+
+
+def _reraise_gs_errors(uri: str, e: Exception):
+    try:
+        from google.api_core.exceptions import GatewayTimeout, NotFound
+
+        print('Reraising exception: {e.message}')
+
+        # If it's a google service NotFound error
+        if isinstance(e, NotFound):
+            raise FileNotFoundError(f'Object {uri} not found.') from e
+
+        # All clienterror (HTTP 4xx) responses
+        elif isinstance(e, GatewayTimeout):
+            raise ValueError(f'Time out when uploading/downloading {uri} using google cloud storage') from e
+
+    except ImportError as e:
+        raise MissingConditionalImportError(conda_package='google-cloud-storage',
+                                            extra_deps_group='google-cloud-storage',
+                                            conda_channel='conda-forge') from e
+    # Otherwise just raise the original error.
+    raise e
+
+
+class GsObjectStore(ObjectStore):
```

Comment:
Could you please also add a `list_objects` method? I recently added this to other object stores.